expr/func_mock_test.go

```diff
@@ -20,6 +20,9 @@ func (s FuncMock) Context(context Context) Context {
 }
 
 func (s FuncMock) Exec(dataMap DataMap) ([]models.Series, error) {
+	// func_get would retrieve this from the map (added at a higher layer)
+	// Mock should add `data` to properly mock the render path
+	dataMap.Add(Req{}, s.data...)
```

Comment:
this doesn't seem like the right place.
if we call this repeatedly, we will add the same data to datamap repeatedly, which is a contract violation.
adding to datamap should probably be handled in the caller

Reply:
Fair, but the caller for Mock is always tests, which would add to the code duplication. Perhaps `Mock` should own it's own DataMap and `NewMock` should add `in` to the datamap just on creation? This would actually reduce duplication a little, since each function wouldn't need to create its own datamap

Reply:
@Dieterbe - what do you think of the above?

Reply:
i think the current version of this PR is too confusing because it's too subtley different from how the main codepath works. I think your suggestion is a slight improvement but i'm still not convinced it's worth making the testing code so different in architecture/flow just to optimize for LOC.
I think my preferred approach is to have the testing datamap work similarly to the real one, I doubt the extra LOC needed to manually add into the datamap first is an issue.

Reply:
It's not about LOC, it's about how painful it is to change all the tests if we decide to update how datamap works/is instantiated etc.

Reply:
Also, it helps enforce that all tests are doing the same (correct) thing

Reply:
OK then, let's give this a shot:

> Perhaps Mock should own it's own DataMap and NewMock should add in to the datamap just on creation?

Reply:
I'll look into if there is a good way to maintain what we have today without messing with Mock. Maybe another standalone function to init the datamap instead of the `make` line we use now?

Reply:
I think you are right that mock is not a good place for owning the datamap, since some tests need to create multiple Mocks (if they can take multiple seriesLists)